components/google_gemini/actions/common/generate-content.mjs

```diff
@@ -20,6 +21,94 @@ export default {
         }),
       ],
     },
+    text: {
+      propDefinition: [
+        app,
+        "text",
+      ],
+    },
+    responseFormat: {
+      propDefinition: [
+        app,
+        "responseFormat",
+      ],
+    },
+    history: {
+      type: "string[]",
+      label: "Conversation History",
+      description: "Previous messages in the conversation. Each item must be a valid JSON string with `text` and `role` (either `user` or `model`). Example: `{\"text\": \"Hello\", \"role\": \"user\"}`",
+      optional: true,
+    },
+    safetySettings: {
+      type: "string[]",
+      label: "Safety Settings",
+      description: "Configure content filtering for different harm categories. Each item must be a valid JSON string with `category` (one of: `HARASSMENT`, `HATE_SPEECH`, `SEXUALLY_EXPLICIT`, `DANGEROUS`, `CIVIC`) and `threshold` (one of: `BLOCK_NONE`, `BLOCK_ONLY_HIGH`, `BLOCK_MEDIUM_AND_ABOVE`, `BLOCK_LOW_AND_ABOVE`). Example: `{\"category\": \"HARASSMENT\", \"threshold\": \"BLOCK_MEDIUM_AND_ABOVE\"}`",
+      optional: true,
+    },
+  },
+  methods: {
+    formatHistoryToContent(history) {
+      if (!history?.length) {
+        return [];
+      }
+
+      return history.map((itemStr) => {
+        let item;
+        try {
+          item = JSON.parse(itemStr);
+        } catch (e) {
+          throw new ConfigurationError(`Invalid JSON in history item: ${itemStr}`);
+        }
```

Comment:
We should allow for the case where an item in the array is an object or custom expression.
```suggestion
        let item = itemStr;
        if (typeof item === "string") {
          try {
            item = JSON.parse(itemStr);
          } catch (e) {
            throw new ConfigurationError(`Invalid JSON in history item: ${itemStr}`);
          }
        }
```